File: core/api/golang/lib/bulk_command_execution/v0_bulk_command_api/v0_bulk_commands.go
/*
 *    Copyright 2021 Kurtosis Technologies Inc.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package v0_bulk_command_api

import (
	"encoding/json"
	"github.com/kurtosis-tech/kurtosis-client/golang/kurtosis_core_rpc_api_bindings"
	"github.com/kurtosis-tech/stacktrace"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/emptypb"
)

// ====================================================================================================
//                                   Command Arg Deserialization Visitor
// ====================================================================================================

// Visitor that will be used to deserialize command args into
type cmdArgDeserializingVisitor struct {
	bytesToDeserialize         []byte
	deserializedCommandArgsPtr proto.Message
}

func newCmdArgDeserializingVisitor(bytesToDeserialize []byte) *cmdArgDeserializingVisitor {
	return &cmdArgDeserializingVisitor{bytesToDeserialize: bytesToDeserialize}
}

func (visitor *cmdArgDeserializingVisitor) VisitLoadModule() error {
	args := &kurtosis_core_rpc_api_bindings.LoadModuleArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the module-loading args")
	}
	visitor.deserializedCommandArgsPtr = args
	return nil
}

func (visitor *cmdArgDeserializingVisitor) VisitExecuteModule() error {
	args := &kurtosis_core_rpc_api_bindings.ExecuteModuleArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the module-executing args")
	}
	visitor.deserializedCommandArgsPtr = args
	return nil
}

func (visitor *cmdArgDeserializingVisitor) VisitRegisterService() error {
	args := &kurtosis_core_rpc_api_bindings.RegisterServiceArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the register service args")
	}
	visitor.deserializedCommandArgsPtr = args
	return nil
}

func (visitor *cmdArgDeserializingVisitor) VisitStartService() error {
	args := &kurtosis_core_rpc_api_bindings.StartServiceArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the start service args")
	}
	visitor.deserializedCommandArgsPtr = args
	return nil
}

func (visitor *cmdArgDeserializingVisitor) VisitRemoveService() error {
	args := &kurtosis_core_rpc_api_bindings.RemoveServiceArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the remove service args")
	}
	visitor.deserializedCommandArgsPtr = args
	return nil
}

func (visitor *cmdArgDeserializingVisitor) VisitRepartition() error {
	args := &kurtosis_core_rpc_api_bindings.RepartitionArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the repartition service args")
	}
	visitor.deserializedCommandArgsPtr = args
	return nil
}

func (visitor *cmdArgDeserializingVisitor) VisitExecCommand() error {
	args := &kurtosis_core_rpc_api_bindings.ExecCommandArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the exec command args")
	}
	visitor.deserializedCommandArgsPtr = args
	return nil
}

func (visitor *cmdArgDeserializingVisitor) VisitWaitForHttpGetEndpointAvailability() error {
	args := &kurtosis_core_rpc_api_bindings.WaitForHttpGetEndpointAvailabilityArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the endpoint availability-waiting-http-get args")
	}
	visitor.deserializedCommandArgsPtr = args
	return nil
}

func (visitor *cmdArgDeserializingVisitor) VisitWaitForHttpPostEndpointAvailability() error {
	args := &kurtosis_core_rpc_api_bindings.WaitForHttpPostEndpointAvailabilityArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the endpoint availability-waiting-http-post args")
	}
	visitor.deserializedCommandArgsPtr = args
	return nil
}

func (visitor *cmdArgDeserializingVisitor) VisitExecuteBulkCommands() error {
	args := &kurtosis_core_rpc_api_bindings.ExecuteBulkCommandsArgs{}
	if err := json.Unmarshal(visitor.bytesToDeserialize, args); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the bulk command execution args")
	}
	visitor.deserializedCommandArgsPtr = args
	return nil
}

func (visitor *cmdArgDeserializingVisitor) VisitGetServices() error {
	visitor.deserializedCommandArgsPtr = &emptypb.Empty{}
	return nil
}

func (visitor *cmdArgDeserializingVisitor) VisitGetModules() error {
	visitor.deserializedCommandArgsPtr = &emptypb.Empty{}
	return nil
}

func (visitor *cmdArgDeserializingVisitor) GetDeserializedCommandArgs() proto.Message {
	return visitor.deserializedCommandArgsPtr
}

// ====================================================================================================
//                                        Serializable Command
// ====================================================================================================

// Used for serializing
type V0SerializableCommand struct {
	Type V0CommandType `json:"type"`

	// The only allowed objects here are from the bindings generated from the .proto file
	ArgsPtr proto.Message `json:"args"`
}

// A V0SerializableCommand knows how to deserialize itself, thanks to the "type" tag
func (cmd *V0SerializableCommand) UnmarshalJSON(bytes []byte) error {
	interstitialStruct := struct {
		Type      V0CommandType   `json:"type"`
		ArgsBytes json.RawMessage `json:"args"`
	}{}
	if err := json.Unmarshal(bytes, &interstitialStruct); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing the bytes into a command")
	}

	visitor := newCmdArgDeserializingVisitor(interstitialStruct.ArgsBytes)
	if err := interstitialStruct.Type.AcceptVisitor(visitor); err != nil {
		return stacktrace.Propagate(err, "An error occurred deserializing command with the following JSON:\n%v", string(bytes))
	}

	cmd.Type = interstitialStruct.Type
	cmd.ArgsPtr = visitor.GetDeserializedCommandArgs()

	return nil
}


// ====================================================================================================
//                                   Bulk Commands Package
// ====================================================================================================

type V0BulkCommands struct {
	Commands []V0SerializableCommand `json:"commands"`
}

File: core/api/typescript/src/lib/bulk_command_execution/v0_bulk_command_api/v0_bulk_commands.ts
import { ExecCommandArgs, ExecuteBulkCommandsArgs, ExecuteModuleArgs, LoadModuleArgs, RegisterServiceArgs, RemoveServiceArgs, RepartitionArgs, StartServiceArgs, WaitForHttpGetEndpointAvailabilityArgs, WaitForHttpPostEndpointAvailabilityArgs } from "../../../kurtosis_core_rpc_api_bindings/api_container_service_pb";
import { V0CommandTypeVisitor, V0CommandType } from "./v0_command_types";
import { ok, err, Result } from "neverthrow";
import * as protobuf from "google-protobuf";
import * as google_protobuf_empty_pb from "google-protobuf/google/protobuf/empty_pb";

// ====================================================================================================
//                                   Command Arg Deserialization Visitor
// ====================================================================================================

// Visitor that will be used to deserialize command args into
class CmdArgDeserializingVisitor implements V0CommandTypeVisitor {
    private static safeJsonParse = Result.fromThrowable(JSON.parse, CmdArgDeserializingVisitor.parseUnknownExceptionValueToError);

    private readonly bytesToDeserialize: string;
    private deserializedCommandArgsPtr?: protobuf.Message;

    constructor (bytesToDeserialize: string) {
        this.bytesToDeserialize = bytesToDeserialize;
    }

    public visitLoadModule(): Result<null, Error> {
        const deserializationResult: Result<any, Error> = CmdArgDeserializingVisitor.safeJsonParse(this.bytesToDeserialize);
        if (deserializationResult.isErr()) {
            return err(deserializationResult.error);
        }
        const args: LoadModuleArgs = Object.assign(new LoadModuleArgs(), deserializationResult.value);

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitExecuteModule(): Result<null, Error> {
        const deserializationResult: Result<any, Error> = CmdArgDeserializingVisitor.safeJsonParse(this.bytesToDeserialize);
        if (deserializationResult.isErr()) {
            return err(deserializationResult.error);
        }
        const args: ExecuteModuleArgs = Object.assign(new ExecuteModuleArgs(), deserializationResult.value);

    	this.deserializedCommandArgsPtr = args;
    	return ok(null);
    }

    public visitRegisterService(): Result<null, Error> {
        const deserializationResult: Result<any, Error> = CmdArgDeserializingVisitor.safeJsonParse(this.bytesToDeserialize);
        if (deserializationResult.isErr()) {
            return err(deserializationResult.error);
        }
        const args: RegisterServiceArgs = Object.assign(new RegisterServiceArgs(), deserializationResult.value);

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitStartService(): Result<null, Error> {
        const deserializationResult: Result<any, Error> = CmdArgDeserializingVisitor.safeJsonParse(this.bytesToDeserialize);
        if (deserializationResult.isErr()) {
            return err(deserializationResult.error);
        }
        const args: StartServiceArgs = Object.assign(new StartServiceArgs(), deserializationResult.value);

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitRemoveService(): Result<null, Error> {
        const deserializationResult: Result<any, Error> = CmdArgDeserializingVisitor.safeJsonParse(this.bytesToDeserialize);
        if (deserializationResult.isErr()) {
            return err(deserializationResult.error);
        }
        const args: RemoveServiceArgs = Object.assign(new RemoveServiceArgs(), deserializationResult.value);

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitRepartition(): Result<null, Error> {
        const deserializationResult: Result<any, Error> = CmdArgDeserializingVisitor.safeJsonParse(this.bytesToDeserialize);
        if (deserializationResult.isErr()) {
            return err(deserializationResult.error);
        }
        const args: RepartitionArgs = Object.assign(new RepartitionArgs(), deserializationResult.value);

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitExecCommand(): Result<null, Error> {
        const deserializationResult: Result<any, Error> = CmdArgDeserializingVisitor.safeJsonParse(this.bytesToDeserialize);
        if (deserializationResult.isErr()) {
            return err(deserializationResult.error);
        }
        const args: ExecCommandArgs = Object.assign(new ExecCommandArgs(), deserializationResult.value);

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitWaitForHttpGetEndpointAvailability(): Result<null, Error> {
        const deserializationResult: Result<any, Error> = CmdArgDeserializingVisitor.safeJsonParse(this.bytesToDeserialize);
        if (deserializationResult.isErr()) {
            return err(deserializationResult.error);
        }
        const args: WaitForHttpGetEndpointAvailabilityArgs = Object.assign(new WaitForHttpGetEndpointAvailabilityArgs(), deserializationResult.value);

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitWaitForHttpPostEndpointAvailability(): Result<null, Error> {
        const deserializationResult: Result<any, Error> = CmdArgDeserializingVisitor.safeJsonParse(this.bytesToDeserialize);
        if (deserializationResult.isErr()) {
            return err(deserializationResult.error);
        }
        const args: WaitForHttpPostEndpointAvailabilityArgs = Object.assign(new WaitForHttpPostEndpointAvailabilityArgs(), deserializationResult.value);

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitExecuteBulkCommands(): Result<null, Error> {
        const deserializationResult: Result<any, Error> = CmdArgDeserializingVisitor.safeJsonParse(this.bytesToDeserialize);
        if (deserializationResult.isErr()) {
            return err(deserializationResult.error);
        }
        const args: ExecuteBulkCommandsArgs = Object.assign(new ExecuteBulkCommandsArgs(), deserializationResult.value);

        this.deserializedCommandArgsPtr = args;
        return ok(null);
    }

    public visitGetServices(): Result<null, Error> {
        this.deserializedCommandArgsPtr = new google_protobuf_empty_pb.Empty()
        return ok(null);
    }

    public visitGetModules(): Result<null, Error> {
        this.deserializedCommandArgsPtr = new google_protobuf_empty_pb.Empty()
        return ok(null);
    }

    public getDeserializedCommandArgs(): Result<protobuf.Message, Error> {
        if (!this.deserializedCommandArgsPtr) {
            return err(new Error("Deserialized command args pointer was falsy; this indicates that it was never set through a visitor method, which should never happen"));
        }
        return ok(this.deserializedCommandArgsPtr!);
    }

    private static parseUnknownExceptionValueToError(value: unknown): Error {
        if (value instanceof Error) {
            return value;
        }
        return new Error("Received an unknown exception value that wasn't an error: " + value);
    }
}

// ====================================================================================================
//                                        Serializable Command
// ====================================================================================================

// Used for serializing
export class V0SerializableCommand {
    private readonly type: V0CommandType;
    
    // The only allowed objects here are from the bindings generated from the .proto file
    private readonly argsPtr: protobuf.Message;

    constructor (type: V0CommandType, argsPtr: protobuf.Message) {
        this.type = type;
        this.argsPtr = argsPtr;
    } 
       
    public getType(): V0CommandType {
        return this.type;
    }

    public getArgsPtr(): protobuf.Message {
        return this.argsPtr;
    }

}

// // ====================================================================================================
// //                                   Bulk Commands Package
// // ====================================================================================================

export class V0BulkCommands {
    private readonly commands: V0SerializableCommand[];

    constructor (commands: V0SerializableCommand[]) {
        this.commands = commands;
    }
}
